src/core/meshcop/tcat_agent.cpp

```diff
@@ -187,180 +208,152 @@ void TcatAgent::Disconnected(void)
-    bool extendedPanIdsMatch            = false;
+    uint8_t res = kAccessFlag;
 
-    VerifyOrExit(IsConnected());
-    VerifyOrExit(aCommissionerCommandClassFlags & kAccessFlag);
-
-    if (aDeviceCommandClassFlags & kAccessFlag)
-    {
-        additionalDeviceRequirementMet = true;
-    }
-
-    if (!additionalDeviceRequirementMet && (aDeviceCommandClassFlags & kPskdFlag))
+    for (uint16_t flag = kPskdFlag; flag < kMaxFlag; flag <<= 1)
     {
-        additionalDeviceRequirementMet = mPskdVerified;
-    }
-
-    if (!additionalDeviceRequirementMet && (aDeviceCommandClassFlags & kPskcFlag))
-    {
-        additionalDeviceRequirementMet = mPskcVerified;
-    }
-
-    if (mCommissionerHasNetworkName || mCommissionerHasExtendedPanId)
-    {
-        Dataset::Info datasetInfo;
-        Error         datasetError = kErrorNone;
-
-        if (aDataset == nullptr)
+        if (aFlagsChecked & flag)
         {
-            datasetError = Get<ActiveDatasetManager>().Read(datasetInfo);
-        }
-        else
-        {
-            aDataset->ConvertTo(datasetInfo);
-        }
-
-        if (datasetError == kErrorNone)
-        {
-            if (datasetInfo.IsPresent<Dataset::kNetworkName>() && mCommissionerHasNetworkName &&
-                (datasetInfo.Get<Dataset::kNetworkName>() == mCommissionerNetworkName))
+            switch (flag)
             {
-                networkNamesMatch = true;
-            }
-
-            if (datasetInfo.IsPresent<Dataset::kExtendedPanId>() && mCommissionerHasExtendedPanId &&
-                (datasetInfo.Get<Dataset::kExtendedPanId>() == mCommissionerExtendedPanId))
-            {
-                extendedPanIdsMatch = true;
-            }
-        }
-    }
+            case kPskdFlag:
+                if (mPskdVerified)
+                {
+                    res |= flag;
+                }
+                break;
 
-    if (!networkNamesMatch)
-    {
-        VerifyOrExit((aCommissionerCommandClassFlags & kNetworkNameFlag) == 0);
-    }
-    else if (aDeviceCommandClassFlags & kNetworkNameFlag)
-    {
-        additionalDeviceRequirementMet = true;
-    }
+            case kNetworkNameFlag:
+                if (mCommissionerHasNetworkName)
+                {
+                    if (aDatasetInfo == nullptr)
+                    {
+                        res |= flag;
+                    }
+                    else
+                    {
+                        if (aDatasetInfo->IsPresent<Dataset::kNetworkName>() && mCommissionerHasNetworkName &&
+                            (aDatasetInfo->Get<Dataset::kNetworkName>() == mCommissionerNetworkName))
+                        {
+                            res |= flag;
+                        }
+                    }
+                }
+                else
+                {
+                    if (aDatasetInfo == nullptr)
+                    {
+                        res |= flag;
+                    }
+                }
+                break;
```

Comment:
```suggestion
                if (aDatasetInfo == nullptr)
                {
                    res |= flag;
                }
                else if (mCommissionerHasNetworkName && aDatasetInfo->IsPresent<Dataset::kNetworkName>() && mCommissionerHasNetworkName &&
                            (aDatasetInfo->Get<Dataset::kNetworkName>() == mCommissionerNetworkName))
                {
                    res |= flag;
                }
                break;
```

I think this should result with the same logic?